File: src/Lessons/4-CountingElements/4.2-PermChecker.go
package main

import "fmt"

func main() {

	fmt.Println(solution([]int{4, 5, 2, 3}))
	//expect 0
}

func solution(someArray []int) int {
	//Check if array is a permutation
	//Calculate the total sum of ints if permutation
	maxInt := len(someArray)
	expectedSum := ((1 + maxInt) * maxInt) / 2 //Formulae for series sum of  1 + n + ...n+1
	permTracker := make([]bool, maxInt)

	currentSum := 0
	for index := 0; index < len(someArray); index++ {
		num := someArray[index]
		if num > maxInt {
			// a correct permutation will not have a number larger than the largest expected number
			break
		}
		if !permTracker[num-1] {
			permTracker[num-1] = true
			currentSum += num
		}
		if currentSum > expectedSum {
			break
		}
	}
	if currentSum != expectedSum {
		return 0
	}
	return 1

}

File: src/Lessons/4-CountingElements/4.2-PermChecker.js

function solution(A) {
    //Check to see if Array A is a permutation, ie has all numbers 1 to N
    let array = [...A];
    const maxNum = array.length; //ideally the largest number should be the length of the array
    const expectedSum = ((1 + maxNum) * maxNum) / 2; //If we add all the numbers up in permutation series
    let permsTracker = new Array(maxNum).fill(true); //Create an array of similar in length
    
    let sum = 0; // we work to get the sum
   
    for (let index = 0; index < array.length; index++) {
        const num = array[index];
        if(permsTracker[num-1]){
            permsTracker[num-1] = false;
            sum += num;
        }
        //if array contains duplicate or large numbers outside the permutation range
        //The sum will be higher  and automatically fail
        if (sum > expectedSum) 
            break;
    }
    if(sum != expectedSum) return 0;
    return 1;
}
const testArray = [1, 4, 2,3];
console.log(solution(testArray))
//expect 7